File: sdk/typescript/src/session.ts
/**
 * Session persistence for Claude Sentient SDK.
 */

import * as fs from "fs";
import * as path from "path";
import type { SessionState, Phase, Task, GateResult } from "./types";

export class SessionManager {
  private stateDir: string;
  private stateFile: string;
  private historyDir: string;

  constructor(stateDir: string = ".claude/state") {
    this.stateDir = stateDir;
    this.stateFile = path.join(stateDir, "session.json");
    this.historyDir = path.join(stateDir, "history");

    // Ensure directories exist
    fs.mkdirSync(this.stateDir, { recursive: true });
    fs.mkdirSync(this.historyDir, { recursive: true });
  }

  /** Save current session state */
  save(state: SessionState): void {
    state.lastUpdated = new Date().toISOString();
    fs.writeFileSync(this.stateFile, JSON.stringify(state, null, 2));
  }

  /** Load current session state */
  load(): SessionState | null {
    if (!fs.existsSync(this.stateFile)) {
      return null;
    }
    try {
      const data = fs.readFileSync(this.stateFile, "utf-8");
      return JSON.parse(data) as SessionState;
    } catch {
      return null;
    }
  }

  /** Load just the session ID for resume */
  loadSessionId(): string | null {
    const state = this.load();
    return state?.sessionId ?? null;
  }

  /** Clear current session (archive to history first) */
  clear(): void {
    if (fs.existsSync(this.stateFile)) {
      const state = this.load();
      if (state) {
        const archiveFile = path.join(
          this.historyDir,
          `${state.sessionId}.json`
        );
        fs.copyFileSync(this.stateFile, archiveFile);
      }
      fs.unlinkSync(this.stateFile);
    }
  }

  /** Create a new session state */
  create(sessionId: string, profile: string, task: string): SessionState {
    const now = new Date().toISOString();
    const state: SessionState = {
      sessionId,
      startedAt: now,
      lastUpdated: now,
      phase: "init",
      iteration: 1,
      profile,
      task,
      tasks: [],
      gates: {},
      commits: [],
      fileChanges: [],
      metadata: {},
    };
    this.save(state);
    return state;
  }

  /** Update the current phase */
  updatePhase(phase: Phase): void {
    const state = this.load();
    if (state) {
      state.phase = phase;
      this.save(state);
    }
  }

  /** Increment the iteration counter */
  incrementIteration(): void {
    const state = this.load();
    if (state) {
      state.iteration += 1;
      this.save(state);
    }
  }

  /** Record a commit */
  addCommit(commitHash: string): void {
    const state = this.load();
    if (state) {
      state.commits.push(commitHash);
      this.save(state);
    }
  }

  /** Record a file change */
  addFileChange(filePath: string): void {
    const state = this.load();
    if (state && !state.fileChanges.includes(filePath)) {
      state.fileChanges.push(filePath);
      this.save(state);
    }
  }

  /** Update a gate result */
  updateGate(gateName: string, result: GateResult): void {
    const state = this.load();
    if (state) {
      state.gates[gateName] = result;
      this.save(state);
    }
  }

  /** List archived sessions */
  listHistory(): Array<{
    sessionId: string;
    task: string;
    startedAt: string;
    lastUpdated: string;
    profile: string;
  }> {
    const sessions: Array<{
      sessionId: string;
      task: string;
      startedAt: string;
      lastUpdated: string;
      profile: string;
    }> = [];

    if (!fs.existsSync(this.historyDir)) {
      return sessions;
    }

    const files = fs.readdirSync(this.historyDir);
    for (const file of files) {
      if (!file.endsWith(".json")) continue;
      try {
        const data = fs.readFileSync(
          path.join(this.historyDir, file),
          "utf-8"
        );
        const state = JSON.parse(data) as SessionState;
        sessions.push({
          sessionId: state.sessionId,
          task: state.task,
          startedAt: state.startedAt,
          lastUpdated: state.lastUpdated,
          profile: state.profile,
        });
      } catch {
        continue;
      }
    }

    return sessions.sort(
      (a, b) =>
        new Date(b.startedAt).getTime() - new Date(a.startedAt).getTime()
    );
  }
}

export type { SessionState };

File: sdk/python/claude_sentient/session.py
"""Session persistence for Claude Sentient SDK."""

import json
from dataclasses import dataclass, asdict, field
from datetime import datetime
from pathlib import Path
from typing import Any


@dataclass
class SessionState:
    """Persistent session state."""

    session_id: str
    started_at: str
    last_updated: str
    phase: str  # init, understand, plan, execute, verify, commit, evaluate
    iteration: int
    profile: str
    task: str  # The original task/prompt
    tasks: list[dict[str, Any]] = field(default_factory=list)
    gates: dict[str, dict[str, Any]] = field(default_factory=dict)
    commits: list[str] = field(default_factory=list)
    file_changes: list[str] = field(default_factory=list)
    metadata: dict[str, Any] = field(default_factory=dict)


class SessionManager:
    """Persist and restore session state."""

    def __init__(self, state_dir: str | Path = ".claude/state"):
        self.state_dir = Path(state_dir)
        self.state_dir.mkdir(parents=True, exist_ok=True)
        self.state_file = self.state_dir / "session.json"
        self.history_dir = self.state_dir / "history"
        self.history_dir.mkdir(exist_ok=True)

    def save(self, state: SessionState) -> None:
        """Save current session state."""
        state.last_updated = datetime.now().isoformat()
        self.state_file.write_text(json.dumps(asdict(state), indent=2))

    def load(self) -> SessionState | None:
        """Load current session state."""
        if not self.state_file.exists():
            return None
        try:
            data = json.loads(self.state_file.read_text())
            return SessionState(**data)
        except (json.JSONDecodeError, TypeError):
            return None

    def load_session_id(self) -> str | None:
        """Load just the session ID for resume."""
        state = self.load()
        return state.session_id if state else None

    def clear(self) -> None:
        """Clear current session (archive to history first)."""
        if self.state_file.exists():
            # Archive to history
            state = self.load()
            if state:
                archive_file = (
                    self.history_dir / f"{state.session_id}.json"
                )
                archive_file.write_text(self.state_file.read_text())
            self.state_file.unlink()

    def create(
        self,
        session_id: str,
        profile: str,
        task: str,
    ) -> SessionState:
        """Create a new session state."""
        now = datetime.now().isoformat()
        state = SessionState(
            session_id=session_id,
            started_at=now,
            last_updated=now,
            phase="init",
            iteration=1,
            profile=profile,
            task=task,
        )
        self.save(state)
        return state

    def update_phase(self, phase: str) -> None:
        """Update the current phase."""
        state = self.load()
        if state:
            state.phase = phase
            self.save(state)

    def increment_iteration(self) -> None:
        """Increment the iteration counter."""
        state = self.load()
        if state:
            state.iteration += 1
            self.save(state)

    def add_commit(self, commit_hash: str) -> None:
        """Record a commit."""
        state = self.load()
        if state:
            state.commits.append(commit_hash)
            self.save(state)

    def add_file_change(self, file_path: str) -> None:
        """Record a file change."""
        state = self.load()
        if state:
            if file_path not in state.file_changes:
                state.file_changes.append(file_path)
                self.save(state)

    def update_gate(self, gate_name: str, result: dict[str, Any]) -> None:
        """Update a gate result."""
        state = self.load()
        if state:
            state.gates[gate_name] = result
            self.save(state)

    def list_history(self) -> list[dict[str, Any]]:
        """List archived sessions."""
        sessions = []
        for file in self.history_dir.glob("*.json"):
            try:
                data = json.loads(file.read_text())
                sessions.append({
                    "session_id": data.get("session_id"),
                    "task": data.get("task"),
                    "started_at": data.get("started_at"),
                    "last_updated": data.get("last_updated"),
                    "profile": data.get("profile"),
                })
            except (json.JSONDecodeError, KeyError):
                continue
        return sorted(sessions, key=lambda x: x.get("started_at", ""), reverse=True)
